File: pubsubexamples/ex.go
// pubsub.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type Broker struct {
	subscribers map[string][]chan float64
}

func NewBroker() *Broker {
	return &Broker{
		subscribers: make(map[string][]chan float64),
	}
}

func (b *Broker) Subscribe(topic string) <-chan float64 {
	ch := make(chan float64)
	b.subscribers[topic] = append(b.subscribers[topic], ch)
	return ch
}

func (b *Broker) Publish(topic string, msg float64) {
	for _, ch := range b.subscribers[topic] {
		go func(c chan float64) {
			c <- msg
		}(ch)
	}
}

func temperatureSensor(b *Broker, topic string) {
	for {
		temp := 20 + rand.Float64()*5
		fmt.Printf("[Sensor] Publishing: %.2f°C\n", temp)
		b.Publish(topic, temp)
		time.Sleep(2 * time.Second)
	}
}

func hvacSystem(ch <-chan float64) {
	for temp := range ch {
		fmt.Printf("[HVAC] Received: %.2f°C\n", temp)
	}
}

func main() {
	broker := NewBroker()
	topic := "home/temperature/livingroom"

	sub := broker.Subscribe(topic)
	go hvacSystem(sub)
	go temperatureSensor(broker, topic)

	select {} // Keep the main goroutine alive
}

File: pubsubexamples/ex.py
# pubsub.py
import threading
import time
import random
from queue import Queue


class PubSubBroker:
    def __init__(self):
        self.subscribers = {}

    def subscribe(self, topic, callback):
        if topic not in self.subscribers:
            self.subscribers[topic] = []
        self.subscribers[topic].append(callback)

    def publish(self, topic, message):
        if topic in self.subscribers:
            for callback in self.subscribers[topic]:
                callback(message)


# Simulated subscriber
def hvac_system(message):
    print(f"[HVAC] Received temperature: {message}°C")


# Simulated publisher
def temperature_sensor(broker, topic):
    while True:
        temp = round(random.uniform(20.0, 25.0), 2)
        print(f"[Sensor] Publishing temperature: {temp}°C")
        broker.publish(topic, temp)
        time.sleep(2)


if __name__ == "__main__":
    broker = PubSubBroker()
    topic = "home/temperature/livingroom"

    broker.subscribe(topic, hvac_system)

    publisher_thread = threading.Thread(target=temperature_sensor, args=(broker, topic))
    publisher_thread.start()
